### DISCUSSION TITLE
Save Favorite Function with fetched data

### DISCUSSION THREAD

**Discussion by Robert-Schurig** [neuefische/questions#379] [Web]
Hi!
I want to create a feature that allows to save images from an API after clicking a button. These images should be visible  on another page. 

I committed my current code [here](https://github.com/Robert-Schurig/Capstone-App/pull/10) 

On entering the favorites page, there is an error message. There seems to be a problem with the fetching of data in the favorite page. 
This is the error message:
![Bildschirm­foto 2022-12-07 um 13 47 08](https://user-images.githubusercontent.com/114500583/206183190-53e361b7-ce54-4b35-b78f-9254b79bca37.png)

I researched the web for proposals and also asked other students but did not find a solution. It appears to me that the challenge emerges from the combination of saving selected images from one page to another one with using an API. I'd appreciate any help.

**Comment by actyvystom** [comment]
Hi Robert, 
please post also relevant code so that we can get a picture of what might go wrong...
Thanks!

**Reply by Robert-Schurig** [reply]
This is the code for the function that should be called when the button is clicked:

```js

 const [favorite, setFavorite] = useState([]);

  function addFavorite(index) {
    if (favorite.includes(index)) {
      const newFavorite = favorite.filter(favIndex => favIndex !== index);
      setFavorite(newFavorite);
    } else {
      setFavorite([...favorite, index]);
    }
  }

```


And the button with the function:

```js

 <Button onClick={() => addFavorite(favorite)}>
          <Image src={likeIcon} alt="Like Icon" height={80} width={80} />
        </Button>
```
This the code that should render the favorite images:

```js
import {useState, useEffect} from "react";
import Image from "next/legacy/image";

export default function Favorites(favorite) {
  const [favImages, setFavImages] = useState([]);

  useEffect(() => {
    if (favorite.length === 0) {
      setFavImages([]);
    } else {
      fetch(
        "https://api.europeana.eu/record/v2/search.json?media=true&profile=minimal&query=Painting&theme=art&reusability=open&qf=IMAGE_COLOUR:%22TRUE%22&rows=100&wskey=KEY"
      )
        .then(response => response.json())
        .then(favImages => setFavImages(favImages));
    }
  }, [favorite]);

  return (
    <div>
      {favImages.map(favImage => {
        return (
          <article key={favImage.index}>
            <p>{favImage.title}</p>
            <Image
              key={favImage.index}
              src={favImage.edmPreview[0]}
              alt={favImage.title[0]}
              height={300}
              width={500}
            />
          </article>
        );
      })}
    </div>
  );
}

```

**Reply by actyvystom** [reply]
1. You are using two scoped state variables for handling your favorites, so the state variable `favImages` in your `Favorites`-Page will always be detached from the state variable `favorite` you use in your ImageCard-Component... you need to **uplift your state** so that you can pass it down to your components and manage it globally. There are quite a few solutions mentioned here in [discussions using `zustand`](https://github.com/orgs/neuefische/discussions?discussions_q=zustand), but  you could also try to use the built-in [useContext hook of React](https://beta.reactjs.org/apis/react/useContext) 
2. Your `addFavorite`-function never updates your `favorite` state variable because you are passing the (by default) empty state variable onClick into your `addFavorite`-function - so your if-statement will always evaluate to `false`...
